vmdb/app/models/ems_refresh/parsers/openstack_infra.rb

```diff
@@ -0,0 +1,257 @@
+module EmsRefresh
+  module Parsers
+    class OpenstackInfra < Infra
+      def self.ems_inv_to_hashes(ems, options = nil)
+        new(ems, options).ems_inv_to_hashes
+      end
+
+      def initialize(ems, _options = nil)
+        @ems               = ems
+        @connection        = ems.connect
+        @data              = {}
+        @data_index        = {}
+        @host_hash_by_name = {}
+
+        @known_flavors = Set.new
+
+        @os_handle                  = ems.openstack_handle
+        @compute_service            = @connection # for consistency
+        @baremetal_service          = @os_handle.detect_baremetal_service
+        @baremetal_service_name     = @os_handle.baremetal_service_name
+        @orchestration_service      = @os_handle.detect_orchestration_service
+        @orchestration_service_name = @os_handle.orchestration_service_name
+      end
+
+      def ems_inv_to_hashes
+        log_header = "MIQ(#{self.class.name}.#{__method__}) Collecting data" \
+                     " for EMS name: [#{@ems.name}] id: [#{@ems.id}]"
+        $fog_log.info("#{log_header}...")
+
+        load_hosts
+        load_orchestration_stacks
+
+        $fog_log.info("#{log_header}...Complete")
+        @data
+      end
+
+      private
+
+      def servers
+        @servers ||= @connection.servers_for_accessible_tenants
+      end
+
+      def hosts
+        @hosts ||= @baremetal_service.nodes.details
+      end
+
+      def hosts_ports
+        @hosts_ports ||= @baremetal_service.ports.details
+      end
+
+      def stacks
+        @stacks ||= detailed_stacks
+      end
+
+      def detailed_stacks
+        detailed_stacks = []
+        # Seems like Heat doesn't suport detailed list, so we need to call get for each stack, which is suboptimal
+        @orchestration_service.stacks.each { |stack| detailed_stacks << stack.details }
+        detailed_stacks
+      end
+
+      def resources
+        return @resources if @resources
+
+        resources = []
+        stacks.each do |stack|
+          # Nested depth 50 just for sure, although nobody should nest templates that much
+          all_stack_resources = @orchestration_service.list_resources(stack, :nested_depth => 50)
```

Comment:
Hm. In OpenstackInfra templates we do have a lot of nested stacks.

It makes sense to model the nested stacks(resources) as stacks, unless AWS expects, you are able to run the same operations/actions on nested stacks(I really don't know AWS).  You can run actions only on root stack (from stack-list).

Otherwise resource and stack is very similar in Heat, it should have all the same relations, like events, outputs, template (not sure about parameters, might be that is only in root stack). Though not all of them supports the nested-depth arg (only resource list supports it afaik). That would mean several API calls for each nested stack.

Given the complexity, I would probably start with everything flat. Then adding another PR focused on nested stacks. Than another PR with re-factoring the same parts with AWS. Does it sound reasonable?